File: bioinformatics/melt_1cr.r
# melt curve data and fluorescence within 1C range

melt_1cr <- function(floor_temp, 
                     db_usr, db_pwd, db_host, db_port, db_name, 
                     exp_id, stage_id, calib_info, 
                     dye_in='FAM', dyes_2bfild=NULL, 
                     dcv=TRUE, # logical, whether to perform multi-channel deconvolution
                     max_temp=1000.1, 
                     mc_plot=FALSE, 
                     show_running_time=FALSE,
                     ... # options to pass onto `mc_tm_pw`
                     ) {
    
    # start counting for running time
    func_name <- 'melt_1cr'
    start_time <- proc.time()[['elapsed']]
    
    mc_out <- process_mc(db_usr, db_pwd, db_host, db_port, db_name, # for connecting to MySQL database
                         exp_id, stage_id, calib_info, # for selecting data to analyze
                         dye_in, dyes_2bfild, 
                         dcv, max_temp, mc_plot, extra_output=TRUE, show_running_time, ...)
    
    mc_tm <- lapply(mc_out[['mc_bywell']], 
                    function(channel_ele) lapply(channel_ele, 
                                                 function(well_ele) well_ele[['tm']]))
    
    # For each well, average the calibrated fluorescence values for the temperatures 72-73C
    mc_cols <- colnames(mc_out[['fc_wT']][[1]])
    temp_cols <- mc_cols[grepl('temp', mc_cols)]
    fluo_cols <- mc_cols[grepl('fluo', mc_cols)]
    if (length(temp_cols) != length(fluo_cols)) {
        stop('Number of temperature columns is not equal to number of fluorescence columns.') }
    
    fluo_1cr <- lapply(mc_out[['fc_wT']], function(channel_ele) {
        tempsl_1cr_pch <- alply(channel_ele[,temp_cols], .margins=2, 
            .fun=function(temps_pw) {
                tempsl <- temps_pw >= floor_temp & temps_pw < floor_temp + 1
                tempsl[is.na(tempsl)] <- FALSE
                return(tempsl)
                })
        fluo_1cr_pch <- sapply(1:length(fluo_cols), 
                           function(i) mean(channel_ele[tempsl_1cr_pch[[i]], fluo_cols[i]]))
        names(fluo_1cr_pch) <- fluo_cols
        return(fluo_1cr_pch)
        })
    
    mc_w1cr <- list('mc_tm'=mc_tm, '1cr_fluorescence'=fluo_1cr, 'num_channels'=mc_out[['num_channels']]) # each element is a list whose each element represents a channel
    
    # report time cost for this function
    end_time <- proc.time()[['elapsed']]
    if (show_running_time) message('`', func_name, '` took ', round(end_time - start_time, 2), ' seconds.')
    
    return(mc_w1cr)
}

File: bioinformatics/julia/QpcrAnalysis/Essentials/melt_1cr.jl
# melt curve data and fluorescence within 1C range

function melt_1cr(
    floor_tmprtr::Real,
    # copy over the whole signature of `process_mc` and comment out `out_format`
    db_conn::MySQL.MySQLHandle,
    exp_id::Integer, stage_id::Integer,
    calib_info::Union{Integer,OrderedDict};
    # start: arguments that might be passed by upstream code
    well_nums::AbstractVector=[],
    auto_span_smooth::Bool=false,
    span_smooth_default::Real=0.015,
    span_smooth_factor::Real=7.2,
    # end: arguments that might be passed by upstream code
    dye_in::AbstractString="FAM", dyes_2bfild::AbstractVector=[],
    dcv::Bool=true, # logical, whether to perform multi-channel deconvolution
	max_tmprtr::Real=1000, # maximum temperature to analyze
    # out_format::AbstractString="json", # "full", "pre_json", "json"
    verbose::Bool=false,
    kwdict_mc_tm_pw::OrderedDict=OrderedDict() # keyword arguments passed onto `mc_tm_pw`
)

    mc_out = process_mc(
        db_conn,
        exp_id, stage_id, calib_info;
        well_nums=well_nums,
        auto_span_smooth=auto_span_smooth,
        span_smooth_default=span_smooth_default,
        span_smooth_factor=span_smooth_factor,
        dye_in=dye_in, dyes_2bfild=dyes_2bfild,
        dcv=dcv,
        max_tmprtr=max_tmprtr,
        out_format="full", # extra output
        verbose=verbose,
        kwdict_mc_tm_pw=kwdict_mc_tm_pw
    )

    channels = mc_out["channels"]
    fluo_well_nums = mc_out["fluo_well_nums"]
    num_fluo_wells, num_channels = size(mc_out["mc_bywell"])

    mc_tm = [
        mc_out["mc_bychwl"][well_i, channel_i]["Ta_fltd"]
        for well_i in 1:num_fluo_wells, channel_i in 1:num_channels
    ] # an array of the same dimension as `mc_out["mc_bywell"]`

    # For each well, average the calibrated fluorescence values for the temperatures between `floor_tmprtr` and `floor_tmprtr + 1`
    fluo_1cr = [
        begin
            tf = mc_out["tf_bychwl"][channel][well_i]
            tmprtrs_1cr = find(tf["tmprtrs"]) do tmprtr
                floor_tmprtr <= tmprtr < floor_tmprtr + 1
            end # do tmprtr
            mean(tf["fluos"][tmprtrs_1cr])
        end # begin
        for well_i in 1:num_fluo_wells, channel in channels
    ]

    mc_w1cr = OrderedDict(
        "mc_tm"=>mc_tm,
        "fluo_1cr"=>fluo_1cr,
        "channels"=>channels,
        "fluo_well_nums"=>fluo_well_nums
    )

    return mc_w1cr
end
